In [ ]:
################################
# RUN THIS CELL
################################
from IPython.core.display import HTML
HTML("style.css")

<div class='header_mint'>

# <img style="float: left; padding-right: 10px; width: 60px" src="https://raw.githubusercontent.com/Harvard-IACS/2018-CS109A/master/content/styles/iacs.png"> AC295/CS287/E-115B: Deep Learning for NLP

<br/>
<hr color=black>

## Homework 4: Great Party Trick - Second Edition
### THE MINT BOOK

**Harvard University**<br/>
**Fall 2021**<br/>
**Instructor**: Chris Tanner<br/>
**Release Date**: Oct 19 (Tues) @ 11:59pm<br/>
<font color="red">**Due Date**: Nov 2 (Tues) @ 11:59pm</font>

<hr color=black>
<center>
<div class='quote'>

_"The goal will be to make this one short."_

    Chris Tanner (October 17, 2021)
</div>
</center>
</div>

<div class='header_mint'>
    
# LEARNING OBJECTIVES

</div>
<br/>

The purpose of this homework is to help you:

- think critically about serious ethical issues
- gain practical experience using Transformers for generation
- understand the power and shortcomings of GPT-2
- become more comfortable reading research papers

To assist you reach these learning objectives, this homework is structured into three parts:
- <span style="background-color: #FDFFB6"><b>Foundation (concepts):</b></span> demonstrate an understanding of the core concepts taught in lectures
- <span style="background-color: #FFC8C8"><b>Application (programming):</b></span> gain experience putting that knowledge into practice 
- <span style="background-color: #CAFFBF"><b>Research (creating new knowledge):</b></span> use your current NLP knowledge and skills to go beyond the course material, to grasp cutting-edge results and to critically accept or challenge that information. This serves as practice for you to research your own NLP interests and to be well-equipped to continuously learn the latest, greatest NLP work.

<div class='header_mint'>
    
## SUPPORT

</div>

- **Supplemental Resources:** See the list of [supplemental resources](https://harvard-iacs.github.io/CS287/supplemental) for a wealth of rich information concerning Machine Learning, NLP, and Math. Some of the courses listed concern the exact topics covered in this homework and lectures.
- **Sanity Check cells:** We provide several 'sanity check' cells which allow you to see our expected outputs. You should ensure your code produces the same. <span style="background-color: #FDFFB6"><b>**NOTE:** We are not claiming that passing the sanity check cells indicates that you have _fully_ implemented everything correctly; rather, they provide simple checks to help inform you if you are on the right track.</span>
- **Ed**: If you are stuck on anything conceptual (not code) about the content from lectures, please post a question on Ed. This is your community, and please contribute and help each other out. If your questions concern the homework, you can post these on Ed, too, but make sure you are not posting any of your code or solutions in general. If you think you've spotted a bug in our homework questions, or something that needs clarifying, please let us know on Ed! We want to correct these issues ASAP.
- **OH:** After having given a wholehearted attempt, if you are having trouble with the homework, please come to Office Hours.
- **Classmates:** We have a strict policy about the homeworks being individual. You are free to discuss _concepts_ with one another, to help each other learn the material. However, no student shall ever discuss their solutions or see another student's solutions to any problem. Once you see someone's coding solution, it's nearly impossible to harness that information in a way that you can write your own unique solution. You've been robbed of a learning opportunity and will likely just regurgitate someone else's work. As a reminder, if you want to take a shortcut on any problem by looking online for already-existing solutions, that's permissible, but you must cite your sources. Otherwise, it constitutes cheating. Posting any pieces of this homework online for others to see if a flagrant violation of our academic policy.
- **Other:** I want everyone to be and feel fully supported. If there's anything else we, as a teaching staff, can do to further assist in your learning, please let us know. Related, at the end of this homework assignment, you are expected to complete an anonymous feedback form. I urge you to critically and earnestly think about your own learning, communicate to us your thoughts, and to optionally tell us possible adjustments we could make so that you meet our learning expectations and you achieve your own learning goals.

<div class='header_purple'>
    
# 1. FOUNDATION (CONCEPTS) [8 points]

</div>
<br/>

<div class='q_purple'><b>1.1 Debiasing [2 points]</b>

In Lecture 13, Ellie Lasater-Guttmann gave a lecture on biased embeddings, where we mostly discussed type-based word embeddings. At the end, she mentioned that contextualized word representations (aka token-based embeddings) are not immune, as they can lead to undesirable consequences, too. In general, which do you think would be harder to debias, contextualized word representations (i.e., token-based) or word embeddings (i.e., type-based)? Explain your reasoning in 3-5 sentences.
</div>

## My Answer:

I think it is harder to debias contextualized word representations than word embeddings. In particular, I believe that in many cases the bias that appear when using certain words in certain situations are represented not only by the word itself but more by the context in which the word is used. In this sense, the contextualized word representations capture better the context in which a word is used and thus these are the embeddings/vectors that will capture the information from the context as well as the information from the word itself. In this sense, the approach we had of modifying the corpus for debiasing our representations might be harder to achieve since we would not only need to change the word itself but rather change the whole context of the word.  

<div class='q_purple'><b>1.2 Responsibility [6 points]</b>

Large language models (LLMs) like BERT and GPT-2 are usually developed by big software companies who have the computational resources and interest to create such. Keep in mind that LLMs will continue to be developed by a myriad of companies for the foreseeable future (for decades, I'd guess), so this is not a dig at the creators of BERT (Google) or GPT-2 (OpenAI). Let's say you've just graduated from Harvard and are working for a company named Transparent AI (I'm making up this name, so if such a company exists, it's not intentional). You are on a team of 50 people who work on building a LLM, and your specific job is to make an aspect of the model more parallelizable and computationally efficient.

**Q1 (3 points)**: Do you feel you have any responsibility to help ensure the model's predictions are equitable and as bias-free as possible? If so, then what are possible actions you should/would take? To be clear, this question isn't asking what technical aspects you would add to the model; it's asking what human-level actions you would take, such as speaking with others, volunteering to do extra engineering work, etc. Please discuss with a minimum of 3 sentences.

**Q2 (3 points)**: The company is large and very hierarchical. Everyone has quarterly expectations, a manager to appease, yet a finite amount of time and personal lives and hobbies outside of work. Where should responsibility lie in addressing these issues? Meaning, who all should be accountable? Please discuss with a minimum of 3 sentences.
</div>

## My answer

Q1: I do have a responsibility to help ensure the model's predictions are equitable and as bias-free as possible. I could probably set up recurrent meetings with other developers to make sure we are all doing something to make our part to have a bias-free model and pplan methods to continually test if our embeddings are starting to show significant bias. I could also volunteer to get bias-free datasets or convert portions of our training datasets to be bias free. I could also try to get partnership with institutes/universities that do work to prevent bias in AI to do workshops/get projects going that prevent the predictions to be biased.

Q2: I think mostly the managers should receive the responsibility of making sure that the predictions are unbiased. I say this because they are the ones responsible of managing the team and as such they should make sure that the work needed to prevent biases in our predictions is distributed among the team members. IN that sense, everyone in the team will be responsible for some small piece of this task and as such the whole team will be responsible for making sure the predictions are not biased.

<div class='header_pink'>
    
# 2. APPLICATION (PROGRAMMING) [60 points]

</div>


# Fine-Tuning GPT2 for Song Lyric Generation

You've made it to the last programming part! We promise this one will be short(ish).

In this section, we'll experiment with OpenAI's GPT2 model in a song lyric generation task. Like BERT, GPT2 is a *large language model* (LLM), and has
achieved historically stunning results on a wide variety of NLP tasks.
As in the last homework, because training GPT2 from scratch is 
infeasible on your laptop, we will borrow the pretrained GPT2 weights published on [HuggingFace](https://huggingface.co/transformers/model_doc/gpt2.html) and finetune them for our downstream task.

You will:
1. Pick your favorite artists (either from our provided dataset, or feel free to bring your own!)
2. Finetune GPT2 on your artists' lyrics
3. Generate novel lyrics

As before, *we strongly recommend you use Colab for this assignment*. Unless you're one of the lucky few who've managed to snag an RTX 3080 amidst this silicon shortage (or have a similarly beefy GPU on hand), **you will not be able to run this notebook on your local machine.**

**DISCLAIMER**: We offer the dataset of lyrics on an as-is basis for pedagogical purposes. We do not support, condone, or affiliate with any problematic messages potentially embedded in this dataset. To view the original source, see [here](https://www.kaggle.com/deepshah16/song-lyrics-dataset).

*A side note*: you've probably heard of [GPT3](https://en.wikipedia.org/wiki/GPT-3), OpenAI's successor to GPT2. So why aren't we using GPT3? For a couple reasons:
* The model is ginourmous. It's an order of magnitude *bigger* than GPT2. As you will soon see, Colab's GPUs can barely handle the *smallest* GPT2 model. If we gave you GPT3, we'd need a dedicated datacenter just for this class.
* The model is **closed source**. In a rather controversial move, OpenAI awarded Microsoft [exclusive access](https://www.technologyreview.com/2020/09/23/1008729/openai-is-giving-microsoft-exclusive-access-to-its-gpt-3-language-model/) to the model. Researchers can still use a public API to query to the model, but only Microsoft has access to the weights directly.

In the past, OpenAI has cited social concerns around releasing GPT3 openly to the public, claiming that GPT3-generated text can be difficult to distinguish from human-authored text. Malicious individuals could, for example, use it to generate large quantities of fake news. Hence, even prior to the Microsoft deal, OpenAI has never openly released the training weights.

All that's to say, we won't be able to use GPT3 in the assignment. But if you're curious to learn more, do check out [the original paper](https://arxiv.org/abs/2005.14165). Also, if you're a fan of text-based dungeon crawlers, take a look at [AI Dungeon](https://play.aidungeon.io/main/home), which uses GPT3 to drive their
"Dragon AI" (the free tier uses GPT2).

All right, enough chat, let's dive in! (after a brief warning message)


**WARNING**: GPT2 is *huge*. We will use the smallest model in the GPT2 family, but even so, it will stretch Colab to breaking. More so than in previous homeworks, we will have to pay attention to memory usage. For tips on writing
memory-efficient PyTorch code, check out their [FAQ](https://pytorch.org/docs/stable/notes/faq.html) and their blerb on [memory management](https://pytorch.org/docs/stable/notes/cuda.html#cuda-memory-management).

Some tools that may come in handy for debugging memory issues:
* [torch.cuda.memory_allocated()](https://pytorch.org/docs/stable/generated/torch.cuda.memory_allocated.html#torch.cuda.memory_allocated)
* [torch.cuda.max_memory_allocated()](https://pytorch.org/docs/stable/generated/torch.cuda.max_memory_allocated.html)
* [gc.collect()](https://docs.python.org/3/library/gc.html)
* [torch.cuda.empty_cache()](https://pytorch.org/docs/stable/generated/torch.cuda.empty_cache.html)

You can find the size (in bytes) of any tensor using
```python
t = make_huge_pytorch_tensor()
bytes_per_element = t.element_size()
total_elements = t.nelement()

total_size = bytes_per_element * total_elements
```

One final note, whenever a cell fails to execute, temporary variables declared
in that cell may not always be deleted by Python. To handle a memory leak, you can try to force a garbage collection cycle and release any cached GPU memory:

```python
import gc  # Python's garbage collection interface

gc.collect()
torch.cuda.empty_cache()
```

However, any variables still in reference somewhere in the notebook will not be
garbage-collected. The easist way to refresh the entire slate is to just
restart your kernel and run everything from the top.

**The important takeaway**: If you get an error that looks like

```
RuntimeError: CUDA out of memory
```
The easiest (sometimes only) way out of this is to restart the whole notebook and run it from the top. **Simply rerunning your cells without restarting will probably give you another memory error**.

<div class='header_pink'>

## Setting up

</div>

As in the previous homework, let's install tranformers

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
PARENT_DIR = '/content/drive/MyDrive/hw4/' # for google colab. adjust accordingly

In [ ]:
! pip install transformers

Double-check Colab has allocated a GPU for you. If not, remember to select from the top bar `Runtime --> Change runtime type` and set `Hardware Acceleration` to
`GPU`. Then run the following cell to print out diagnostic info about your GPU: 

In [ ]:
! nvidia-smi

Wed Nov  3 02:21:58 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 495.29.05    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla P100-PCIE...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   48C    P0    40W / 250W |      0MiB / 16280MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

Here are some suggested imports for you to use. Feel free to add any additional you would like! (as long as they don't trivialize the assignment, of course)

In [ ]:
from typing import List

import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from tqdm import tqdm

import torch
from torch.utils.data import Dataset, DataLoader
from transformers import GPT2TokenizerFast, GPT2Model, get_linear_schedule_with_warmup

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

Because you may need to restart and rerun your notebook many times as you deal with memory issues, you may want to keep your data files on Google drive rather than in Colab's local storage. The local storage is cleared every time Colab allocates a new compute instance for you. To learn more about connecting to Drive, see [this example notebook](https://colab.research.google.com/notebooks/io.ipynb).

<div class='header_pink'>

## Data preparation

</div>

<div class='q_pink'><b>2.1 Assemble the data [3 points]</b>
    
We've pre-processed a lyrics dataset for you. It contains 3 columns: the artist, title, and lyrics. Every artist in the dataset has at least 100 unique songs, and there are 12 unique artists represented. 

For the sake of training speed, we encourage you to choose 4 artists of the 12 to use. Then, from the data, create a series of prompts for GPT-2 by creating a single string per song. Each string should contain, in the following order:
- An `<|endoftext|>` token 
- The tokens `song artist:`
- The name of the artist
- The tokens `song title:`
- The title of the song
- The tokens `song lyrics`:
- The lyrics of the song
- An `<|endoftext|>` token 

This is so that we can use this formatting as a prompt for GPT-2 to generate titles and lyrics later, conditioned on artist.

Then, split your dataset into a train and validation set. We recommend using 80-90% for training.
    
</div>

In [ ]:
# Read in, process, and split your data below
df = pd.read_csv(f'{PARENT_DIR}combined_lyrics.csv')
print(df.shape)
sub_df = df[df['Artist'].isin(['Eminem', 'Lady Gaga', 'Beyoncé', 'Maroon 5'])]
sub_df.iloc[:100,:]

(2491, 3)


Artist  ...                                              Lyric
0   Eminem  ...  look if you had one shot or one opportunity to...
1   Eminem  ...  rihanna i'm friends with the monster that's un...
2   Eminem  ...  ed sheeran i've been a liar been a thief been ...
3   Eminem  ...  i got a song filled with shit for the strongwi...
4   Eminem  ...  eminem now this shit's about to kick off this ...
..     ...  ...                                                ...
95  Eminem  ...  bizarre music reality sometimes it's hard to t...
96  Eminem  ...  verse some days i just wanna up and call it qu...
97  Eminem  ...  a lot of people ask me where the fuck i've bee...
98  Eminem  ...  produced by bass brothers   baby your dada lov...
99  Eminem  ...  in my mind i'm a fighter my heart's a lighter ...

[100 rows x 3 columns]

In [ ]:
EOT_TOKEN = '<|endoftext|>'

def string_definer(text):
  output_string = EOT_TOKEN
  output_string += ' song artist: '
  output_string += text['Artist']
  output_string += ' song title: '
  output_string += text['Title']
  output_string += ' song lyrics: '
  output_string += text['Lyric']
  output_string += ' ' + EOT_TOKEN
  return output_string

concatenated_text = sub_df.apply(string_definer, axis=1)
concatenated_text

0      <|endoftext|> song artist: Eminem song title: ...
1      <|endoftext|> song artist: Eminem song title: ...
2      <|endoftext|> song artist: Eminem song title: ...
3      <|endoftext|> song artist: Eminem song title: ...
4      <|endoftext|> song artist: Eminem song title: ...
                             ...                        
705    <|endoftext|> song artist: Maroon 5 song title...
706    <|endoftext|> song artist: Maroon 5 song title...
707    <|endoftext|> song artist: Maroon 5 song title...
708    <|endoftext|> song artist: Maroon 5 song title...
709    <|endoftext|> song artist: Maroon 5 song title...
Length: 710, dtype: object

In [ ]:
from sklearn.model_selection import train_test_split

train_text, test_text  = train_test_split(
    concatenated_text.tolist(), train_size=0.90, test_size=0.10
)
print(f'Training Size: {len(train_text)}')
print(f'Testing Size: {len(test_text)}')

Training Size: 639
Testing Size: 71


<div class='q_pink'><b>2.2 Make a SongLyrics Dataset [3 points]</b>

Using your split  data, prepare a SongLyrics dataset to use for training.

It's up to you how you'd like to structure your dataset, as long as it works with your model and training code further down. Some suggestions:
* Use [GPT2TokenizerFast](https://huggingface.co/transformers/model_doc/gpt2.html#gpt2tokenizerfast), a speedier variant of the tokenizer, written in Rust
* `GPT2TokenizerFast.from_pretrained('gpt2')` will load the GPT2 (small version) weights from HuggingFace. The [docs for the parent class `PreTrainedTokenizer`](https://huggingface.co/transformers/v2.11.0/main_classes/tokenizer.html#pretrainedtokenizer) may also be helpful. 
* By default GPT2 does not include padding tokens. For reasons that will soon be clear, we won't really need them in either in this assignment, but if you'd like, feel free to add your own padding token with `tokenizer.add_special_tokens()`
* When tokenizing, ensure that your examples are shorter than the max sequence
length accepted by GPT2 (1024 tokens). An easy way to assure this is to set `truncation=True` when calling your tokenizer. See [the docs](https://huggingface.co/transformers/internal/tokenization_utils.html#transformers.tokenization_utils_base.PreTrainedTokenizerBase) for a full and exhaustive list of all the useful customizations you can try
    
</div>

In [ ]:
class SongLyrics(Dataset):
    def __init__(self, data: List[str]):
        #raise NotImplementedError   # TODO: implement        pass
        self.tokenizer = GPT2TokenizerFast.from_pretrained('gpt2')
        self.tokenized_strings = self.tokenizer(data, truncation=True)['input_ids']
        self.vocab_size = self.tokenizer.vocab_size

    def __len__(self):
        #raise NotImplementedError   # TODO: implement        pass
        return len(self.tokenized_strings)
    def __getitem__(self, idx: int):
        #raise NotImplementedError   # TODO: implement        pass
        return torch.tensor(self.tokenized_strings[idx])

In [ ]:
tokenizer = GPT2TokenizerFast.from_pretrained('gpt2')

In [ ]:
tokenizer.vocab_size

50257

In [ ]:
train_ds = SongLyrics(train_text)
test_ds = SongLyrics(test_text)

In [ ]:
print(len(train_ds))
print(len(test_ds))
print(train_ds[0])
print(test_ds[0])
print(train_text[0])
print(test_text[0])

639
71
tensor([50256,  3496,  6802,    25,  1526,  2049,   642,  3496,  3670,    25,
        40011,    72,  3496, 15844,    25,   345,   260,   884,   257, 44109,
         1312,   760,   345,  5938,   290,   523,   466,  1312,  1312, 48004,
         1096,  1312,   766,   345,   503,   345,  1239, 19951,   257,  5273,
          326,   356,  1239,  4888,   475,   663,   523,  6283,   663,  1223,
          649,  4998,  7666,   326,  1312,   423,   329,   345,  1312,  1969,
          616,  2951,   618,   545,  3436,  4240,   644,   340,    67,   307,
          588,   284,   787,   345, 44567,   220,   662,  1312, 18869,  1577,
          345,  1223,  1365,   621,  1997,   345,   303,  1683,   550,   257,
         7387,   290,   257,  5443, 18854,   262,   995,   340, 27934,   523,
         3049,   220,   220,  6029,   479, 14246,    72,   534, 35919, 37472,
          866,   616, 22531,   523,  3387,  1309,   502, 17666,  2245,   340,
          878,   340,  6140,   220,   220,   523,  1577, 

<div class='q_pink'><b>2.3 Dataloaders [2 points]</b>

Make new dataloaders out of your dataset. For the dataloader, you may find the option `pin_memory=True` to be helpful, as it caches tensors in GPU memory for quicker access. See [this thread](https://discuss.pytorch.org/t/when-to-set-pin-memory-to-true/19723) for details.

**VERY IMPORTANT**: On both of your dataloaders, set `batch_size=1`. Any larger of a batch_size and **you will run into memory issues**.
</div>

In [ ]:
train_dl = DataLoader(train_ds, batch_size=1, shuffle=True, drop_last=True, pin_memory=True)
test_dl = DataLoader(test_ds, batch_size=1, shuffle=False, drop_last=True, pin_memory=True)
#raise NotImplementedError   # TODO: implement

<div class='header_pink'>

## Model and training

</div>

To access the GPT2 weights, we will use HuggingFace's GPT2Model class. Please take a moment to familiarize yourself with the class's [API](https://huggingface.co/transformers/model_doc/gpt2.html#gpt2model), especially the `forward()` method and its return type.

We can instantiate a GPT2Model like so:

In [ ]:
model = GPT2Model.from_pretrained('gpt2')

This gives you a PyTorch module that you can use just like any other. To feed an input through the model, simply do:

In [ ]:
example = next(iter(train_dl))
output = model(example)

We can inspect the shape of our output:

In [ ]:
print(example.shape)
print(output[0].shape)

torch.Size([1, 723])
torch.Size([1, 723, 768])


In [ ]:
flatten_output = torch.squeeze(output[0])#output[0].view(output[0].shape[1], output[0].shape[2])#torch.flatten(output[0],2)
flatten_output.shape

torch.Size([723, 768])

In [ ]:
example

tensor([[50256,  3496,  6802,    25, 11182, 42192,  3496,  3670,    25, 45328,
         12860,  3496, 15844,    25,   705, 25587,  1312,  9149,   345,  1053,
          1239,  1775,   257,   905,   588,   428,   878,  8066,   905,   345,
          1223,   326,   345,   460,   470,  8856, 18869,  1011,   257, 27185,
           788,  1280,   510,   262, 41160,   705, 25587,  1312,  9149,   345,
          1053,  1239,  1775,   257,   905,   588,   428,   878,  8066,   905,
           345,  1223,   326,   345,   460,   470,  8856, 18869,  1011,   257,
         27185,   788,  1280,   510,   262, 41160,  5156,   340,   338,   257,
         22601, 12860,   220,   220,  1312, 18869,   760,   644,   466,   345,
           588,   644,   561,   345,   910,   611,  1312,   531,  1312,  1549,
           466,  1997,   284,   345,   345,   765,  1312, 18869,   760, 10194,
           340,   338, 23036,   649,  6116,   731,   345,   561,   588,   345,
          1053,   587,  8196,   259,     6,  2245,  

What do you think each of those dimensions corresponds to? (a rhetorical question, no need to type an answer here! But please do think about this before reading on)

```
 _________________________________________
( Hmmm... what do those dimensions means? )
 -----------------------------------------
        o   ^__^
         o  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||
```

...

...

...

All right so the dimensions are

$$ B \times N \times H$$
where:
* $B$ is batch size
* $N$ is sequence length
* $H$ is hidden size

As the documentation indicates, `last_hidden_state` corresponds to the output of the final decoder block in GPT2, constructing an embedded representation of each of the $N$ inputs. Each embedded representation is a vector of length $H$. The (small) GPT2 model uses $H = 768$.



<div class='q_pink'><b>2.4 GPT2 Language Model [16 points]</b>

Below, you will find skeleton code outlining the shape of a GPT2 language model. Using the base GPT2Model class, we will construct on top a language model head that predicts the next word after the current input. For example, suppose our sentence is:

```
The quick brown fox jumps over the lazy
```

From this sentence, the GPT2Model constructs a sequence of 8 embeddings

$$ h_1, h_2, ..., h_8$$
where

$$ h_1  = f(The)$$
$$ h_2 = f(The\; quick)$$
$$ h_3 = f(The\; quick\; brown)$$

That's to say, the $i$th embedding is constructed as a function over the previous $i$ inputs. Then given the embedding $h_i$, we'd like to predict the $i+1$st word. So our end output should be 8 words

$$ w_1, w_2, w_3, ..., w_8$$

where

$$ w_1 = g(h_1)$$
$$ w_2 = g(h_2)$$
$$ w_3 = g(h_3)$$

And ideally, after training we should have that 

$$w_1 \approx quick$$
$$w_2 \approx brown$$
$$w_3 \approx fox$$

All the way until $w_8 \approx dog$ (or whatever mammal the fox was feeling like jumping over today).

So to recap, we have the weights from the base model GPT2Model, which produces a hidden vector for each token in the input sequence. Our goal is to use GPT2Model to build a language model that achieves the above prediction task. The only component missing is some way to transform the GPT2 embeddings into predicted words, then use a loss function to compare the predictions to the actual next words. As in HW2, pay close attention to the alignments between input words and predicted words, so as to ensure that your model isn't simply predicting the next word.

**CAUTION:** While you have the freedom to design this as you'd like, take the easy route and be frugal where you can! Your sequence length can be up to 1024 tokens, and the vocabulary size is over 50,000. Even though we're using a batch size of only 1, keep in mind the sheer size (in MB) of the tensors at play, not to mention the gradients PyTorch is saving along the way!
   
</div>

In [ ]:

def flatten_tensors(gpt_output: torch.tensor) -> torch.tensor:
    return gpt_output.view(gpt_output[1], gpt_output[2])

class GPT2LanguageModel(torch.nn.Module):
    def __init__(self, output_vocab_size: int):
        super().__init__()
        self.hidden_size = 768 # parameters fixed by GPT2
        self.model = GPT2Model.from_pretrained('gpt2')
        self.linear = torch.nn.Linear(self.hidden_size, 512) #output_vocab_size)
        self.dropout = torch.nn.Dropout(p=0.2)
        self.linear2 = torch.nn.Linear(512, output_vocab_size)

        
    def forward(self, input_ids):
        gpt_result = self.model(input_ids)
        output = self.linear(gpt_result[0])
        output = self.dropout(output)
        output = self.linear2(output)
        return output

def compute_loss(preds, target_ids):
    preds = preds[:,:-1,:]
    target_ids = target_ids[:, 1:]
    preds_squeeze = torch.squeeze(preds,0)
    target_ids_squeeze = torch.squeeze(target_ids,0)
    return torch.nn.CrossEntropyLoss()(preds_squeeze, target_ids_squeeze)



<div class='q_pink'><b>2.5 Train the model [10 points]</b>

It's time to train the beast! **Train your model below, printing out train and validation loss at regular intervals.**

We (and the [Huggingface team](https://github.com/huggingface/transformers/blob/master/src/transformers/training_args.py)) recommend using `Adam` or `AdamW` as an optimizer with a learning rate of `5e-5`, no weight decay, and default beta and epsilon parameters (`betas = [0.9, 0.999]`, `eps = 1e-8`). You may need to tweak these parameters slightly to achieve good model performance. 

You may want to consider using a [linear scheduler](https://huggingface.co/transformers/main_classes/optimizer_schedules.html#transformers.get_linear_schedule_with_warmup) for the learning rate. By default, Huggingface uses a linear scheduler with no warmup and which takes a scheduler step for every optimizer step (instead of every epoch).

Finally, since we are forced to use a batch size of 1 due to memory constraints, you may also want to consider using [gradient accumulation](http://kozodoi.me/python/deep%20learning/pytorch/tutorial/2021/02/19/gradient-accumulation.html). In PyTorch, this just means doing multiple `loss.backward()` across batches before calling `optimizer.step()` and `optimizer.zero_grad()`.

If everything goes to plan, you should see times of around 5 minutes per epoch, depending on the size of your datasets. Training for 5 - 10 epochs is plenty (you could probably get away with even fewer).

To take advantage of GPU acceleration, as in the previous homework, remember to move your model to the GPU:

```
model.cuda()
```

as well as your training examples as you receive them:

```
example_tensor = example_tensor.cuda()
```

**MEGA CAUTION**: If you haven't read any of our previous warnings yet, read this one!

When you move your model and examples to the GPU, you will start dipping into the very finite pool of VRAM (video RAM) available on Colab's cut-rate GPUs. If you've followed our previous advice (batch_size=1, frugal model), you should be fine. However, the moment you run this cell, **your model and examples will stay in GPU memory**. Worse, if your cell fails with an error or you halt execution prematurely, your cell may leak memory.

For example, Colab has probably allocated for you a Tesla K80 GPU with 12 GB of VRAM. For us, running the training cell will take almost 7GB of VRAM. Once this cell finishes (or throws an error), that 7GB of VRAM continues hanging around. That means the next time we run this cell, if that memory hasn't been released (and it probably hasn't), PyTorch will try to allocate *another 7GB*, but because there isn't another 7GB free, it will return a `CUDA out of memory` error. To manually free that memory, you can try some of our tips from above (manual garbage collection, deleting unused variables, emptying the cache), but it probably won't get everything. 

**If all else fails, the easiest (sometimes only) way to empty your memory is to restart your kernel and rerun the notebook from the start.**

To do so, on the top bar select `Runtime --> Restart and run all`.
    
</div>


In [ ]:
# implement your training loop here
def train(vocab_size: int, train_dl: DataLoader, val_dl: DataLoader):
    model = None
    losses = {
        'train': [], # keep track of your losses in these lists
        'val': []
    }
    
    model = GPT2LanguageModel(vocab_size)# Seq2Seq(train_dl.dataset.source_vocab_size, train_dl.dataset.target_vocab_size, 512, PADDING_IDX)
    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
    model.to(device)
    
    #loss_function = nn.CrossEntropyLoss() #NLLLoss()
    optimizer = torch.optim.Adam(model.parameters(), lr=0.00005)#, weight_decay=1e-5)
    #scheduler = get_linear_schedule_with_warmup(optimizer, 2000, 10) #worked well 200 and 1200

    epochs = 6
    for epoch in range(epochs):
        train_running_loss = 0.0
        for i, train_batch in enumerate(train_dl):
            train_batch = train_batch.to(device)

            train_preds = model(train_batch)
            train_targets = train_batch

            train_loss = compute_loss(train_preds, train_targets)
            
            optimizer.zero_grad()
            train_loss.backward()
            optimizer.step()
            
            train_running_loss += train_loss.item()
        
        #scheduler.step()
        
        print('Epoch %d, training_loss %.3f'% 
            (epoch + 1, train_running_loss / len(train_dl))) 
        
        losses['train'].append(train_running_loss/len(train_dl))
        
        val_running_loss = 0.0
        for i, val_batch in enumerate(val_dl):
            with torch.no_grad():
                val_batch = val_batch.to(device)

                val_preds = model(val_batch)
                val_targets = val_batch

                val_loss = compute_loss(val_preds, val_targets)

                val_running_loss += val_loss.item() 

        print('Epoch %d, validation_loss %.3f'% 
            (epoch + 1, val_running_loss / len(val_dl))) 
        
        losses['val'].append(val_running_loss/len(val_dl))
        
    print("Finished training!")
    
    return model, losses


In [ ]:
model, losses = train(tokenizer.vocab_size, train_dl, test_dl)


Epoch 1, training_loss 6.208
Epoch 1, validation_loss 4.719
Epoch 2, training_loss 4.325
Epoch 2, validation_loss 4.098
Epoch 3, training_loss 3.741
Epoch 3, validation_loss 3.812
Epoch 4, training_loss 3.309
Epoch 4, validation_loss 3.614
Epoch 5, training_loss 2.937
Epoch 5, validation_loss 3.529
Epoch 6, training_loss 2.576
Epoch 6, validation_loss 3.486
Finished training!


<div class='header_pink'>

## Evaluation

</div>


<div class='q_pink'><b>2.6 Generate text [6 points]</b>
    
Now the fun part, playing with our trained model! To start, fill out the method below that generates a novel sequence of text. Given some primer text, you should use GPT2 to continually generate the next token until it hits either the max sequence length or the end of text token `<|endoftext|>`.

Some hints:
* Observe that with a single invocation, GPT2 will only predict one additional new token. The other $N - 1$ predictions are for tokens in your input sequence. Thus, you will have to call GPT2 many times, feeding back the primer plus the new predicted word, to produce a sequence of novel words.
* To generate different examples with each call, rather than calling `argmax` over your word probabilities, why not use them as sampling probabilities instead? That is, if your vocabulary is `[ball, cow, planet, grass]` and your probabilities for the next word are `[0.1, 0.5, 0.2, 0.2]`, then with probability $0.5$ the next word should be `cow`.
* To convert an index back into the original token, use HuggingFace's `tokenizer.decode()` method. Check out the [API](https://huggingface.co/transformers/internal/tokenization_utils.html#transformers.tokenization_utils_base.PreTrainedTokenizerBase) for details.
    
</div>

In [ ]:
import torch.nn.functional as F

tokenizer = GPT2TokenizerFast.from_pretrained('gpt2')

@torch.no_grad()
def generate(primer: str, max_length=50):
    tokenized_primer = torch.tensor(tokenizer(primer, truncation=True)['input_ids'])
    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
    tokenized_primer = tokenized_primer.to(device)
    next_word_preds = model(tokenized_primer)[-1,:]

    # Temperature sampling
    temp = 2
    next_word_preds = F.softmax(next_word_preds/temp, dim=0)
    # Top k 
    k = 15
    sorted_vals, _ = torch.sort(next_word_preds)#[-20]
    top_k_cut = sorted_vals[-20]
    zero_tensor = torch.zeros(next_word_preds.shape[0]).to(device)
    next_word_preds_k = torch.where(next_word_preds > top_k_cut, next_word_preds, zero_tensor)

    next_token = torch.multinomial(next_word_preds_k, 1).item()
    next_word = tokenizer.decode(next_token)

    j = len(primer.split(' '))
    
    del tokenized_primer

    #print("This is j ", j)
    while j <= max_length:
        #print("have not found eot_token")
        #print("This is primer ", primer)
        if next_word == EOT_TOKEN:
            #print("Found end of text token")
            break
        primer = primer + ' ' + next_word
        tokenized_primer = torch.tensor(tokenizer(primer, truncation=True)['input_ids'])
        tokenized_primer = tokenized_primer.to(device)
        next_word_preds = model(tokenized_primer)[-1,:]

        # Temperature sampling
        #temp = 0.8
        next_word_preds = F.softmax(next_word_preds/temp, dim=0)
        # Top k 
        #k = 7
        sorted_vals, _ = torch.sort(next_word_preds)#[-20]
        top_k_cut = sorted_vals[-20]
        zero_tensor = torch.zeros(next_word_preds.shape[0]).to(device)
        next_word_preds_k = torch.where(next_word_preds > top_k_cut, next_word_preds, zero_tensor)

        next_token = torch.multinomial(next_word_preds_k, 1).item()
        next_word = tokenizer.decode(next_token)  
        j += 1
        del tokenized_primer

    return primer

    #import pdb
    #pdb.set_trace()

Try running the cell below to test your `generate()` function. Feel free to play around with different primers to get a feel for how your model behaves.

In [ ]:

n_examples = 5
max_length = 20
primer = "i wonder if" #"your prompt here"
for i in range(n_examples):
    ex = generate(primer, max_length)
    print(f'Ex {i}: {ex}')

Ex 0: i wonder if  saw est er o oh  yeah  uh  o oo oho  yeah  4  o oh  i  only  like  one
Ex 1: i wonder if  artist  best  world  way  that  way  you  could  see    when  you  gotta  know    you  are  between
Ex 2: i wonder if  much  feel  again  there  is  where  they 're  along  from  hell  be  gonna    until  he    
Ex 3: i wonder if  artist  real  part  hook    if  i  feel  hard  as  when  i  look  when  that  girl  your  girls
Ex 4: i wonder if  good  eyes  good  eyes  if  in  a  girl  who  she  d  again    your  face  and  that  life



<div class='q_pink'><b>2.7 Unconditioned Generation [5 points]</b>
    
    
Now lets do a few particular examples. Generate $10$ songs without conditioning (i.e. with only the `<|endoftext|>` token for a primer). Write a few sentences evaluating how your model performs. 
    
</div>

In [ ]:
n_examples = 10
max_length = 100
primer = EOT_TOKEN
for i in range(n_examples):
    ex = generate(primer, max_length)
    print(f'Ex {i}: {ex}')

Ex 0: <|endoftext|>  song  artist :  song  will    i  won 't  be  too  many  years    we  must  we  will  i  and  i  will      it 's  tough  and  it 's  like  i  can  be  there  before  this  love  it  does  in  our  love  our  words  the  city  up  together  i  i  won 't  come    like  i    did  y  let  we    you  and  yeah    won  ' cause  i  in  us    the  night  of  our    we  both    we    must    you  ' tin  y  yeah    hook  i    let 's  work   i
Ex 1: <|endoftext|>  song :  radio  one  two  i 'm  still  more  people  and  if  you  only  baby  baby    i  oh  baby    and  if  i  love  o oo oh  you    where  how  did  you  fall  a  u  want    you  give  ya  hey  my  style    how  did  i    oh  yeah  yeah    your  life    there  were  i  no  life  without  my    baby  i  dont  wanna  oh    where  in  myself    your  baby  girl  uh  wo ah  u      you  you    if    k im at  all  i  don 't    do  you  
Ex 2: <|endoftext|>  song  artist  moment  t cause  he  don g  know    that  guy  oh

From the unconditional generation of text, we can see that the model learned some of the structure of the data. We can see that the predictions have the 'song artist: ' portion, many of the predictions have some form of the artist name. Additionally, we can see that there si some variety in the words, not all the words are the same and some words that are expected appear such as 'gaga' (from lady gaga songs) or 'eminem (from eminem songs).

<div class='q_pink'><b>2.8 Lyric Completion [5 points]</b>

### Lyric Completion Evaluation

Now let's use the beginning of each song as a primer. For a couple of songs in the validation set, omit the final $\{10\%, 30\%, 50\%, 70\%, 90\%\}$ of each song and generate the remaining lyrics for the song. 

Write a few sentences comparing your predictions to the true lyrics. How do these differ between artists? How does the length of the primer affect the style and accuracy of the completed lyrics?

In [ ]:
import math

iter_obj = iter(test_dl)
song1 = next(iter_obj)
song2 = next(iter_obj)

#import pdb
#pdb.set_trace()

song1_tokens = tokenizer.decode(song1[0])#.split(':')
song2_tokens = tokenizer.decode(song2[0])#.split(':')

split_song1_tokens = song1_tokens.split(':')
split_song2_tokens = song2_tokens.split(':')
print(song1_tokens)
print(song2_tokens)

print("Song 1 artist ", split_song1_tokens[1])
print("Song 1 title ", split_song1_tokens[2])

print("Song 2 artist ", split_song2_tokens[1])
print("Song 2 title ", split_song2_tokens[2])

song1_lyrics = split_song1_tokens[3].split(' ')
num_words_song1 = len(song1_lyrics)

song2_lyrics = split_song2_tokens[3].split(' ')
num_words_song2 = len(song2_lyrics)

percent_to_remove = [0.10, 0.30, 0.50, 0.70, 0.90]

def generate_perc_func(percent_list, song_tokens):
  song_tokens_split = song_tokens.split(':')
  song_lyrics_list = song_tokens_split[3].split(' ')
  #import pdb
  #pdb.set_trace()
  num_words_song = len(song_lyrics_list)
  print(f"Length of song is {num_words_song}")
  for percent in percent_list:
    print(f"Generating last {percent * 100}%")
    subset_lyrics = song_lyrics_list[:-math.floor(percent * num_words_song)]
    print(f"Length of subset lyrics is {len(subset_lyrics)}")
    primer_lyrics = ' '.join(subset_lyrics)
    print(f"Primer lyrics are {primer_lyrics}")
    primer = ':'.join(song_tokens_split[:3]) + ': ' + primer_lyrics
    print(f"Primer length is {len(song_tokens_split[0]) + len(song_tokens_split[1]) + len(song_tokens_split[2]) + len(subset_lyrics)}")
    #import pdb
    #pdb.set_trace()
    generated_lyrics = generate(primer, num_words_song)
    print(f"The length of generated lyrics is {len(generated_lyrics.split(' '))}")
    print(f"Generated last {percent * 100}% lyrics : {generated_lyrics}")
    print("\n")
    print("-"*40)
    print("\n")

print("\n")
print("Generation for song 1")
print("\n")
generate_perc_func(percent_to_remove, song1_tokens)


print("\n")
print("Generation for song 2")
print("\n")
generate_perc_func(percent_to_remove, song2_tokens)



<|endoftext|> song artist: Lady Gaga song title: Bad Romance (Chew Fu H1N1 Fix) song lyrics: walk walk fashion baby work it move that bitch crazy walk walk fashion baby work it move that bitch crazy walk walk fashion baby work it move that bitch crazy walk walk fashion baby work it move that bitch crazy walk walk fashion baby work it move that bitch crazy walk walk fashion baby work it move that bitch crazy walk walk fashion baby work it move that bitch crazy walk walk passion baby work it i'm a free bitch  hook free bitch x9 i'm a free bitch free x5 i'm a free x7 bitch free x5  i'm a free bitch baby  post raraahahah romaromama gaga oh lala want your bad romance raraahahah romaromama gaga oh lala want your bad romance   i want your ugly i want your disease i want your everything as long as its free i want your love love love love i want your love i want your drama the touch of your hand i want your leatherstudded kiss in the sand i want your love love love love i want your love love lo

From the examples above we can see that the pattern is that for the lyrics missing 10% or 30% of the words, the prediction have words that are related to the rest of the song or that appear before. However, as we deleted more of the lyrics (for 50%, 70% and 90%) we see that the predicted words are very different from the actual song lyrics. In particular, we see that many of the predicted words are either blank spaces or very simple, one-syllable words.

<div class='q_pink'><b>2.9 Song Title Generation [5 points]</b>


Now condition on only the artist names and generate several song titles per artist. Write a few sentences about the generated titles. How realistic are these titles? How do they compare to the other titles' of the artist?

In [ ]:
#raise NotImplementedError   # TODO: implement

song_title_per_artist = 3
artist_name = ['Eminem', 'Lady Gaga', 'Beyoncé', 'Maroon 5']

def generate_title(artist_list, num_titles):
    for artist in artist_list:
        print(f"Generating {num_titles} for {artist}")
        for i in range(num_titles):
            primer = EOT_TOKEN + ' song artist: ' + artist  + ' song title: ' 
            ex = generate(primer, 18)
            
            print(f"The title generated is: {ex}")

generate_title(artist_name, song_title_per_artist)

Generating 3 for Eminem
The title generated is: <|endoftext|> song artist: Eminem song title:     dr  and  shady    he  must as  far  as  is  there
The title generated is: <|endoftext|> song artist: Eminem song title:   mark  my  daughter  j im  ready  to  take  it  i 'm ma
The title generated is: <|endoftext|> song artist: Eminem song title:   hook  you  see  my  love  ' cause  that  boy  it 's  all
Generating 3 for Lady Gaga
The title generated is: <|endoftext|> song artist: Lady Gaga song title:   Rap !  You  gotta  call  ya    don 't  buy  me
The title generated is: <|endoftext|> song artist: Lady Gaga song title:   The  Sun �  Night  and  the  air  you  get  i  i
The title generated is: <|endoftext|> song artist: Lady Gaga song title:   With  My - oker  dat  hey  k r igg ar is
Generating 3 for Beyoncé
The title generated is: <|endoftext|> song artist: Beyoncé song title:   speech  o ow  so  many  we  r y ork  how  there  americ
The title generated is: <|endoftext|> song artist: Be

We can see that the results in the title generation are not the best and it is probably related to the fact that there might be some data leakage in our training pipeline that we could not fix. We can see that for songs from Eminem, we can see titles with words that appear more frequently in that author songs, as well as with Lady Gaga. For Beyonce, every time we ran this cell, we found words related to love, baby, rlationships. The hardest to predict song titles were from Maroon 5. For some reason the model had a harder time predicting the title from this band.

<div class='q_pink'><b>2.10 Full Song Generation [5 points]</b>

Now condition on artist names and *new* song titles of your creation. Write a few sentences evaluating the resulting lyrics. How realistic are they given your titles? How reasonable are they for a song in general? How does their style compare to other lyrics by that artist?  


In [ ]:
#raise NotImplementedError   # TODO: implement

artist_name = ['Eminem', 'Lady Gaga', 'Beyoncé', 'Maroon 5']
song_titles = {'Eminem': ['8 mile', 'Da dude', 'WTF your momma'],
               'Lady Gaga': ['Rockester', 'Monster city', 'Crazy Rose'],
               'Beyoncé': ['Why you leave me?', 'Wanna be my lover', 'Remember us'],
               'Maroon 5': ['Sugar', 'Leave me', 'Lets get party']}

def generate_lyric(artist_list, song_titles):
    for artist in artist_list:
        print(f"Generating songs for {artist}")
        for title in song_titles[artist]:
            primer = EOT_TOKEN + ' song artist: ' + artist  + ' song title: ' + title + ' song lyrics: '
            ex = generate(primer, 350)
            print(f"The song generated is: {ex}")

generate_lyric(artist_name, song_titles)

Generating songs for Eminem
The song generated is: <|endoftext|> song artist: Eminem song title: 8 mile song lyrics:   i 'm  back  up  j ork  and  i  want  ' cause  if  you  i  want    '  cause  i        so  baby  wait  and  i  don 't  know  baby  you  just  gotta  go  and    the  man  im  so  alright  but  now  it 's  only  that  i  say    em inem  im a  girl  i  said  he 's  too  late  i  dont  try  this  time    em  i    feel  so  good  to  be  happy  as  long  as    and  yeah  i  don  give    and  im  perfect    you  as  long  as  cause  you    im  a it    im  in  my  middle  finger  up  j  on  our  home    i    just  like    just  like    i   'm  yeah    im  a  i    a  guy    it  just  gotta    me  and  i    ' cause  i      and      ' cause  ' yon a  y or go y d aw g i  like  this  and  yeah    im  on  i  n ah  ' round    em  in em    it  was  too  much  it 's  so  i  just    you  baby  ' cause  it  ' inem ack      a ight    you  so  girl    i      yo  i      em in ac ou ie  your 

In general our predicted songs are not ery good. They are not very realistic given the titles. In aprticular, many of the predicted songs are random words, very few of the lyrics words are related to the song title. At least, some of the lyrics words seem to be plausible they would appear in that artist songs. The style that some of the lyrics seem to posses, in terms of the words that are used, do align with the kind oflyrics that that artist usually sings, especially this was the case for many Eminem and Lady Gaga songs.

<div class='header_green'>
    
# 3. RESEARCH PAPER 1 [16 points]

</div>

**NOTE:** Since I want everyone to dedicate the vast majority of the remaining semester to the research project, I highly suggest selecting papers that are related to your project. And, for this final homework, we will ask you to read two papers, not just one.


-----------

As we build a foundation in NLP, it's also important to also see what the latest, cutting-edge work (research) looks like. It's incredibly worthwhile to learn about the types of problems people work on, their methodology and approach to the problem, the datasets they work on, the issues they raise, and the solutions they posit. The field moves incredibly fast, but the __approach__ to ML/NLP research is relatively stable -- different types of papers are accepted as the years progress, but that's a different story.

We want to help you get practice reading research papers, which mostly entails thinking critically about the work, being able to discern the main takeaways/conclusions, and to reflect on the work in a meaningful way.

<div class='q_green'><b>3.1: Read an NLP research paper [0 points]</b>
    
Select and read a paper that was published in ACL, NAACL, EMNLP, or COLING in 2020 or 2021. You can find a list of such published papers by searching Google, a la "ACL 2020 accepted papers". For this assignment, you are allowed to pick either a short paper (4-5 pages) or long paper (8-9 pages), **but you must not select a workshop paper**. List below the name of the paper, authors, venue, and year published.
</div>

The title of the paper i am reading is "GNN is a counter? Revisiting GNN for question answering" by Kuan Wang, Yuyu Zhang, Diyi Yang, Le Song, Tao Qin. The paper was posted in Arxiv in October 2021, not yet published but done in Microsoft Research and quite relevant to our final project.

<div class='q_green'><b>3.2: Problem? [2 point]</b> What is the problem that it is trying to address? In other words, what is it trying to solve? (2-3 sentences)
</div>

THe paper is trying to answer the questions: Are GNN under- or over-complicating networks for QA? What is the essential role of the graph architecture in reasoning over knowledge?

<div class='q_green'><b>3.3: Solution? [2 point]</b> At a very high-level, what was their solution? (2-3 sentences). Here, you don't have room to go into the small details (e.g., about the model), so you'll need to summarize the most important elements that comprised the solution.
</div>

The authors used Sparse Variational Dropout as a tool to disssec the existing graph network architecture. This tools was used to recognize which pieces of the QA-GNN were dispesable and which ones were actually cotnributing to reasoning. Based on these results of the dissection, the authors came up with a simplified GNN model that shows great results on the CommonSenseQA and OpenBookQA.


<div class='q_green'><b>3.4: Data? [2 points]</b> What dataset(s) did they use? Are they freely available? What's the size of the data? (2-3 sentences)
</div>

The authors used CommonSenseQA and OpenBookQA datasets which are datasets commonly used as benchmark for question answering tasks. Both of the datasets are freely available. CommonSenseQA is a multiple choice question dataset with 12,102 questions and OpenBookQA is a multiple choicd question dataset with 5,957 multiple choice questions. They also use ConceptNet as a general commonsense graph. ConceptNet is also freely available to use and has over 13 million of edges. 

<div class='q_green'><b>3.5: Model [2 points]</b> Very related to the 'solutions' question, describe here any models that they used, and what made it effective (2-3 sentences)
</div>

With the SParse Variational Dropout, the authors found that GNN are over-parameterized, some layers can be pruned substantially and the initial node embeddings are dispensable. After this discover, they developed Graph Soft Counter (GSC) an extremely simple graph neural network that basically serves as a counter of the edges of the knowledge graph. This much simpler model, with less than 1% of trainable parameters than the used GNN architectures in other papers, outperforms the previous solutions. 

<div class='q_green'><b>3.6: Results? [2 points]</b> What are their main results? (~2 sentences)
</div>

The GSC network outperforms the results of other GNN based networks in both benchmark datasets. In particular, GSC outperforms the previous best model with a mean of 2.57% test accuracy in the CommonSenseQA dataset. Similarly, GSC outperforms the previous best model with a mean of 2.53% test accuracy in the OpenBookQA dataset. 

<div class='q_green'><b>3.7: Strengths? [2 points]</b> List 2-3 strengths of the paper
</div>

- The use of SVD (SParse Variational Dropout) is a great idea to show really how much the weights/layers of the GNN are helping with reasoning.
- The simple structure of GSC and great results in test set strongly show that there is an overparameterization of the GNN models when applied to this task.

<div class='q_green'><b>3.8: Weaknesses? [2 points]</b> Although you may be new to this problem and all of its content, try to list 2-3 weaknesses of the paper (anything that you think could strengthen the paper is sufficient).
</div>

- Even though GSC has a clear and intuitive explanation, the detailed explanation in the paper does not seem very clear to me. I think there could be more work in the explanation here.
- Although it was a great idea to include the SVD, it would have been helpful to provide more background on how it was used on the model.

<div class='q_green'><b>3.9: Evaluation [2 points]</b>
    
How would you evaluate this paper in terms of:
- scientific contribution
- effectiveness to solve the problem
- how convincing it was.
    
Give each of these elements a score from 1-10 (10 is best). No word explanation necessary.
</div>

- scientific contribution: 9
- effectiveness to solve the problem: 9
- how convincing it was : 9
    

<div class='q_green'><b>3.10: Research Ideas [4 points]</b>
    
Think of 1-2 research ideas that you have based on this paper. It doesn't have to be grand; most research is very incremental. Specifically, your research idea should have a concrete question that you're aiming to answer. List it below. 
</div>

- The idea to show that GNN is overparameterized is really cool. However, I think it would be interesting to see if such overparameterization holds for other type of question answering, i.e. question answering for non-multiple choice questions. I think this could provide insights into how effective the regular GNN's are for this more general questions and if GSC is enough for this. 

<div class='header_green'>

    
# 4. RESEARCH PAPER 2 [16 points]

</div>

And another one! -- DJ Khaled

<div class='q_green'><b>4.1: Read an NLP research paper [0 points]</b>
    
Select and read a paper that was published in ACL, NAACL, EMNLP, or COLING in 2020 or 2021. You can find a list of such published papers by searching Google, a la "ACL 2020 accepted papers". For this assignment, you are allowed to pick either a short paper (4-5 pages) or long paper (8-9 pages), **but you must not select a workshop paper**. List below the name of the paper, authors, venue, and year published.
</div>

The title of the paper is "Enhancing multiple choice question answering with causal knowledge" written by Dhairya Dalal, Mihael Arcan, and Paul Buitelaar. The paper was published in ACL in 2021.

<div class='q_green'><b>4.2: Problem? [2 point]</b> What is the problem that it is trying to address? In other words, what is it trying to solve? (2-3 sentences)
</div>

The paper explores the effectiveness of augmenting pretrained language models with external causal knowledge for multiple-choice causal question answering. In particular, the authors prsent new strategies for representing the causal knowledge and augenting pretrained models with external causal knowledge.


<div class='q_green'><b>4.3: Solution? [2 point]</b> At a very high-level, what was their solution? (2-3 sentences). Here, you don't have room to go into the small details (e.g., about the model), so you'll need to summarize the most important elements that comprised the solution.
</div>

The paper extracts causal facts from CauseNet dataset, it creates tuples by extracting a lits of tokens $T$ from the input question/premise. Then the authors generate a list of potential causal fact candidates that are filtered once. Finally, they create a table that can be queried to return all candidate facts where both cause and effects exist as tokens. Then the authors present ways to do causal knowledge representations: Distributed Causal Embeddings and Causal Knowledge Graph Embeddings.

<div class='q_green'><b>4.4: Data? [2 points]</b> What dataset(s) did they use? Are they freely available? What's the size of the data? (2-3 sentences)
</div>

The authors use CauseNet that is a causality graph extracted from the web, publicly available with more than 11 million causal relations. Additionally, the authors use the COPA (Choice of Plausible Alternatives) dataset which has more than multiple choice 1000 questions and is publicly available. And, the authors also use the WIQA (What-IF Question Answering) dataset whih has 39706 questions containing a perturbation and a possible effect in the context of a paragraph, it is also publicly available. 

<div class='q_green'><b>4.5: Model [2 points]</b> Very related to the 'solutions' question, describe here any models that they used, and what made it effective (2-3 sentences)
</div>

The authors present causal knowledge representation models to solve their problem. In particular, they describe a method for modelling causality using a distrbituional similarity model. Two embeddings are learned fo cause and effect concepts respectively, where their model learns embeddings fo reach single token representation of caus and effect concepts and finally get one single representation using mean pool between cause and effect vectors. The authors also present a mehtod to represent causal knowledge as a knowledge graph embedding. The authors represent entities and relations in a lower-dimensional space such that the head, the relation and tail satisfy the dimensionality: $h + r = t$. Once we get an embedding for the head, relation and tail, we get a single vector by mean pooling these vectors.

<div class='q_green'><b>4.6: Results? [2 points]</b> What are their main results? (~2 sentences)
</div>

The authors used RoBERTa as a baseline and they get a 4% increase in the COPA dataset using the distributional similarity model and a 6.20% using the causal knowledge graph representation. In the WIQA dataset, the authors also get a 7% incrase using causal knowledge graph represenation but not better results using distributional similarity model. 


<div class='q_green'><b>4.7: Strengths? [2 points]</b> List 2-3 strengths of the paper
</div>

- The paper makes a strong case for leveraging causal knowledge sources for causal question answering.
- I think it is a strong suit of the paper to consider different ways of using the causal knowledge representations. This provide substantial footing to use causal knowledge graphs and GNN in the future.  

<div class='q_green'><b>4.8: Weaknesses? [2 points]</b> Although you may be new to this problem and all of its content, try to list 2-3 weaknesses of the paper (anything that you think could strengthen the paper is sufficient).
</div>

- It is not quite clear, I had to spend quit a bit of time to figure it out, how the authors are forming causal relationships without using the knowledge graph. That could have been explained better.
- From my understanding in the field, and given the recency of the paper, the authors could have used other knowledge graphs that seem to be more established, e.g. ConceptNet.

<div class='q_green'><b>4.9: Evaluation [2 points]</b>
    
How would you evaluate this paper in terms of:
- scientific contribution
- effectiveness to solve the problem
- how convincing it was.
    
Give each of these elements a score from 1-10 (10 is best). No word explanation necessary.
</div>

- scientific contribution: 8
- effectiveness to solve the problem: 7
- how convincing it was: 8

<div class='q_green'><b>4.10: Research Ideas [4 points]</b>
    
Think of 1-2 research ideas that you have based on this paper. It doesn't have to be grand; most research is very incremental. Specifically, your research idea should have a concrete question that you're aiming to answer. List it below. 
</div>

- We could try to test how it affects the source of knowledge graph in answering the questions in a same dataset. IN particular, for COPA and WIQA, we can change the knowledge base fro CauseNet to ConceptNet and see how the results change.

<div class='q_green'><b>BONUS POINTS [5 points]</b> I mention the full details in the syllabus on the course website. However, in short, these bonus points cannot bring one's grade to exceed 100. That is, if someone received a 97 on this homework, doing this bonus could allow their grade to reach 100 points. If the person had an 83 on the homework, then the most they could achieve is an 88.
    
The task: read another research paper -- meaning, a total of 3 papers. It's allowed to be a Short Paper. Write answers to the same questions again. Please copy and paste all of the questions below.
</div>

<div class='q_green'><b>E.1: Read an NLP research paper [0 points]</b>

The last paper I am reading is: "Multi-Scale Progressive Attention Network for Vieo Question Answering" by Zhicheng Guo, Jiaxuan Zhao, Licheng Jiao, Xu Liu, Lingling Li, it was published in ACL in 2021.

<div class='q_green'><b>E.2: Problem? [2 point]</b> What is the problem that it is trying to address? In other words, what is it trying to solve? (2-3 sentences)
</div>

The authros are trying to solve the problem of answering questions from videos. IN particular, they want to answer how can we use video information to answer specific questions related to the video.

<div class='q_green'><b>E.3: Solution? [2 point]</b> At a very high-level, what was their solution? (2-3 sentences). Here, you don't have room to go into the small details (e.g., about the model), so you'll need to summarize the most important elements that comprised the solution.
</div>

The authors proposed a new network Multi-Scale Progressive Attention Network (MSPAN) to achieve relational reasoning between cross-scale video information. In particular, the authors generate clip-level features from the video oand generate a graph from each time step of clips. Then apply GNN to extract features used for answering the question.

<div class='q_green'><b>E.4: Data? [2 points]</b> What dataset(s) did they use? Are they freely available? What's the size of the data? (2-3 sentences)
</div>

The authors used the datasets TGIF_QA which is a large-scale benchmark dataset for VideoQA. This dataset is publicly availabe and contains 165000 QA pairs for animated GIFs. Also, the authors used the MSVD-QA and MSRVTT-QA tasks which are generated from video descriptions.  

<div class='q_green'><b>E.5: Model [2 points]</b> Very related to the 'solutions' question, describe here any models that they used, and what made it effective (2-3 sentences)
</div>

The proposed model has two parts. First the authors obtain video representations using ResNet networks to extract features from the video and to extract video motion features. Also, they generate a question embedding representation using a BiLSTM network. Afterwards, the authors use max-pool layers with different kernel sizes to generate multi-scale graphs. On top of each graphs, the authors run a GCN and combine the results with the question embedding. Finally, they use a decoder to get the answer to the question.

<div class='q_green'><b>E.6: Results? [2 points]</b> What are their main results? (~2 sentences)
</div>

MSPAn produces state of the art methods in terms of action questions, count questions, FrameQA and transitions quesitons on the TGIF_QA dataset. Similarly the gate SOTA resutls fro most groups of questions for the MSVD-QA and MSRVT-QA datasets.

<div class='q_green'><b>E.7: Strengths? [2 points]</b> List 2-3 strengths of the paper
</div>

- The explanation of the network is very clear and although it could be complicated the authors made a great job in putting that clear.
- The authors were able to effectively use GNNs for their network. I think looking at the video scales through the lens of GNNs could open the door for future research in this direction.

<div class='q_green'><b>E.8: Weaknesses? [2 points]</b> Although you may be new to this problem and all of its content, try to list 2-3 weaknesses of the paper (anything that you think could strengthen the paper is sufficient).
</div>

- Although the authors had a good explanation of how to use GNN's in their particular application. It was hard for me to understand why or how they got to the conclusion that they can use these networks.
- The authors could have put an appendix to explain their model in detail. Is a short fault, but I think there could have been space for writing a few more of the math and details about how the model was designed.

<div class='q_green'><b>E.9: Evaluation [2 points]</b>

- scientific contribution: 8
- effectiveness to solve the problem: 8
- how convincing it was: 9

<div class='q_green'><b>E.10: Research Ideas [4 points]</b>
    
Think of 1-2 research ideas that you have based on this paper. It doesn't have to be grand; most research is very incremental. Specifically, your research idea should have a concrete question that you're aiming to answer. List it below. 
</div>

- It would be interesting to see if we can extract richer information and surass the results if we use other type of graph neural network architectures. 

<div class='header_blue'>
    
# 5. SELF-REFLECTION [0 points]

</div>

<div class='q_blue'><b>5.1: Self-reflection and Feedback [0 points]</b>

Are you thriving in the course? Are there elements that are particularly confusing to you? I want everyone to be and feel fully supported. Toward this, I strongly urge you all to think critically about your own learning and efforts. Please provide us with feedback about how you're doing in the course and if there's anything further or different we can do to better assist your learning. I want everyone to give their earnest account, so the form is completely anonymous.

</div>

[Anonymous Self-Reflection and Feedback Form](https://forms.gle/3LT6UfhtCtqp2G7X9)